data/index.js

```diff
@@ -0,0 +1,52 @@
+/**
+ * External dependencies
+ */
+import { createStore } from 'redux';
+import { flowRight } from 'lodash';
+
+/**
+ * Module constants
+ */
+const reducers = [];
+const enhancers = [];
+if ( window.__REDUX_DEVTOOLS_EXTENSION__ ) {
+	enhancers.push( window.__REDUX_DEVTOOLS_EXTENSION__() );
+}
+const store = createStore( dynamicReducer, {}, flowRight( enhancers ) );
+
+/**
+ * Reducer function combining the dynamic "reducers" array
+ *
+ * @param {Object} state  Global state
+ * @param {Object} action Action
+ *
+ * @return {Object}       Updated global state
+ */
+function dynamicReducer( state = {}, action ) {
+	let hasChanges = false;
+	const newState = {};
+	reducers.forEach( ( { key, reducer } ) => {
+		const newSubState = reducer( state[ key ] || {}, action );
+		hasChanges = hasChanges || newSubState !== state[ key ];
+		newState[ key ] = newSubState;
+	} );
+
+	return hasChanges ? newState : state;
+}
+
+/**
+ * Register a new sub reducer to the global sate
+ *
+ * @param {String} key     Reducer key
+ * @param {Object} reducer Reducer function
+ */
+export function registerReducer( key, reducer ) {
```

Comment:
To address the challenges around module-specific middlewares and enhancers, could the data module serve as a registry of stores, rather than of reducers? So each module could have its own middlewares, etc. Main difference would be `subscribe`, `dispatch`, and `getState` would need to operate separately on each store, rather than assuming there's a single store object to work with.

Reply:
Do you think we'd have use-cases requiring cross-module selectors. For example a post-editor module declares a reducer and is dependent on the editor module. In which case, post-editor could provide a cross-module selector? It also seem helpful for plugins.

Granted, you can recreate this with three selectors, one for each store and a third function to merge both results.

---

Also, the other question I have is whether the `data` module should assume a redux store (with middlewares, enhancers) where a `reducer` is not really redux-related. Thinking the `data` module should not be framework-dependent in its external API.

Reply:
I have a few concerns with the registry-of-stores model, as that would break the premise of a single aggregator and pipeline of state for the entire application. If the goal were to provide a common framework for independent plugins to run on, I think it'd make more sense, but here we expect plugins to build on top of one another, and we expect there to be communication to and from Gutenberg, correct?

That said, I think I understand wanting to separate things a little, but to me that would be achieved by keeping enhancements and middlewares domain-specific.

> Also, the other question I have is whether the `data` module should assume a redux store (with middlewares, enhancers) where a `reducer` is not really redux-related. Thinking the `data` module should not be framework-dependent in its external API.

I share this idea, and the broader idea that _learning_ what a reducer is is easier than wrapping one's head around a store.

Reply:
I'm wondering if we could still implement as separate stores, but keep the same top-level interface. i.e.

```js
wp.data.getState = () => stores.reduce( ( result, { store, key } ) => {
    return { ...result, [ key ]: store.getState() };
}, {} );
```

Registering still looks like you're registering a reducer, but you'd have the option to specific middlewares or enhancers if needed.

Might be a nightmare for performance.